sway-core/src/semantic_analysis/cei_pattern_analysis.rs

```diff
@@ -0,0 +1,601 @@
+                    block_name,
+                    warnings,
+                );
+            };
+            set_union(cond_effs, body_effs)
+        }
+        AsmExpression {
+            registers, body, ..
+        } => {
+            let init_exprs = registers
+                .iter()
+                .filter_map(|rdecl| rdecl.initializer.as_ref())
+                .collect();
+            let init_effs = analyze_expressions(init_exprs, block_name, warnings);
+            let asmblock_effs = analyze_asm_block(body, block_name, warnings);
+            if init_effs.contains(&Effect::Interaction) {
+                // TODO: improve locations accuracy
+                warn_after_interaction(&asmblock_effs, &expr.span, &expr.span, block_name, warnings)
+            }
+            set_union(init_effs, asmblock_effs)
+        }
+    }
+}
+
+fn analyze_two_expressions(
+    first: &ty::TyExpression,
+    second: &ty::TyExpression,
+    block_name: &Ident,
+    warnings: &mut Vec<CompileWarning>,
+) -> HashSet<Effect> {
+    let first_effs = analyze_expression(first, block_name, warnings);
+    let second_effs = analyze_expression(second, block_name, warnings);
+    if first_effs.contains(&Effect::Interaction) {
+        warn_after_interaction(
+            &second_effs,
+            &first.span,
+            &second.span,
+            block_name,
+            warnings,
+        )
+    }
+    set_union(first_effs, second_effs)
+}
+
+// Analyze a sequence of expressions
+// TODO: analyze_expressions, analyze_codeblock and analyze_asm_block (see below) are very similar in structure
+//       looks like the algorithm implementation should be generalized
+fn analyze_expressions(
+    expressions: Vec<&ty::TyExpression>,
+    block_name: &Ident,
+    warnings: &mut Vec<CompileWarning>,
+) -> HashSet<Effect> {
+    let mut interaction_span: Span = Span::dummy();
+    let mut accumulated_effects = HashSet::new();
+    let mut analysis_state: CEIAnalysisState = CEIAnalysisState::LookingForInteraction;
+
+    for expr in expressions {
+        let expr_effs = analyze_expression(expr, block_name, warnings);
+        match analysis_state {
+            CEIAnalysisState::LookingForInteraction => {
+                if expr_effs.contains(&Effect::Interaction) {
+                    analysis_state = CEIAnalysisState::LookingForStorageReadOrWrite;
+                    interaction_span = expr.span.clone();
+                }
+            }
+            CEIAnalysisState::LookingForStorageReadOrWrite => warn_after_interaction(
+                &expr_effs,
+                &interaction_span,
+                &expr.span,
+                block_name,
+                warnings,
+            ),
+        };
+        accumulated_effects.extend(expr_effs)
+    }
+    accumulated_effects
+}
+
+// TODO: jumps are not processed at the moment, so the CEI analysis will work only
+// for simple linear assembly blocks, which arguably is the most important use case so far
```

Comment:
Jump are not and will porbably never be allowed in `asm` blocks, so you're good!
```suggestion
No need to worry about jumps because they are not allowed in `asm` blocks.
```